edge/pkg/devicetwin/dmiserver/server.go

```diff
@@ -0,0 +1,185 @@
+package dmiserver
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+	"k8s.io/klog/v2"
+
+	beehiveContext "github.com/kubeedge/beehive/pkg/core/context"
+	beehiveModel "github.com/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/common/constants"
+	pb "github.com/kubeedge/kubeedge/edge/pkg/apis/dmi/v1"
+	messagepkg "github.com/kubeedge/kubeedge/edge/pkg/common/message"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/modules"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/metamanager/dao"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+)
+
+const (
+	SockPath                 string = "/tmp/dmi.sock"
+	UnixNetworkType          string = "unix"
+	ResourceTypeDeviceMapper        = "devicemapper"
+)
+
+type server struct {
+	mapperList      map[string]*pb.MapperInfo
+	deviceList      map[string]*v1alpha2.Device
+	deviceModelList map[string]*v1alpha2.DeviceModel
+}
+
+func (s *server) MapperRegister(ctx context.Context, in *pb.MapperRegisterRequest) (*pb.MapperRegisterResponse, error) {
+	klog.Infof("receive mapper register: %+v", in.Mapper)
+	err := saveMapper(in.Mapper)
+	if err != nil {
+		klog.Errorf("fail to save mapper %s to db with err: %v", in.Mapper.Name, err)
+		return nil, err
+	}
+	s.mapperList[in.Mapper.Name] = in.Mapper
```

Comment:
do we need add lock for the map, it is not safe for multi goroutine

Reply:
i have added a lock for the map

Reply:
here there're 3 maps, and are all from the above global variables, need more considering where to add the lock
```
mapperList      map[string]*pb.MapperInfo
	deviceList      map[string]*v1alpha2.Device
	deviceModelList map[string]*v1alpha2.DeviceModel
```